File: backend/routes/reviews.py
from fastapi import APIRouter, HTTPException
from models.reviews import ReviewRequest, ReviewResponse, ReviewData
from models.base_response import BaseResponse
from models.user import UserRequest
from database.db_instance import db_handler
from uuid import UUID

router = APIRouter(tags=["Reviews"])


@router.get("/me/reviews", response_model=ReviewResponse, status_code=200)
async def get_reviews(request: UserRequest):
    data, err = db_handler.getReview(user_id=request.user_id)
    answer = []
    if data:
        for review in data:
            answer.append(ReviewData(
                rate=review.rate,
                text=review.text,
                book_id=review.book_id,
                user_id=review.user_id,
                created_at=review.created_at
            ))
    return {
        "status": "success",
        "message": "Reviews retrieved",
        "data": answer
    }


@router.get("/reviews/{book_id}", response_model=ReviewResponse, status_code=200)
async def get_review(request: UserRequest, book_id: UUID):
    data, err = db_handler.getReview(user_id=request.user_id, book_id=book_id)
    answer = []
    if data:
        for review in data:
            answer.append(ReviewData(
                rate=review.rate,
                text=review.text,
                book_id=review.book_id,
                user_id=review.user_id,
                created_at=review.created_at
            ))
    return {
        "status": "success",
        "message": "Reviews retrieved",
        "data": answer
    }


@router.put("/me/reviews/{book_id}", response_model=BaseResponse, status_code=200)
async def update_review(request: ReviewRequest, book_id: UUID):
    if not book_id:
        raise HTTPException(status_code=404, detail="Book id not found")
    err = db_handler.updateReview(user_id=request.user_id, book_id=book_id, text=request.text, rate=request.rate)
    return {
        "status": "success",
        "message": "Review updated"
    }


@router.delete("/me/reviews/{book_id}", response_model=BaseResponse, status_code=200)
async def delete_review(request: UserRequest, book_id: UUID):
    if not book_id:
        raise HTTPException(status_code=404, detail="Book id not found")
    err = db_handler.deleteReview(user_id=request.user_id book_id=book_id)
    return {
        "status": "success",
        "message": "Review deleted"
    }


@router.post("/me/reviews/{book_id}", response_model=BaseResponse, status_code=200)
async def create_review(request: ReviewRequest, book_id: UUID):
    if not book_id:
        raise HTTPException(status_code=404, detail="Book id not found")
    err = db_handler.addReview(user_id=request.user_id, book_id=book_id, rate=request.rate, text=request.text)
    return {
        "status": "success",
        "message": "Review created"
    }

File: frontend/src/api/reviews.js
import axios from 'axios'
import config from '@/runtimeConfig'

const api = axios.create({
  baseURL: config.api.baseUrl
})

export const getMyReviews = async (userId) => {
  try {
    console.log('[API] Fetching user reviews...')
    const response = await api.post('/me/reviews', {
      user_id: userId
    })
    
    if (response.data.status !== 'success') {
      const errorMsg = response.data.message || 'Failed to load your reviews. Please try again.'
      console.error('[API] getMyReviews error:', errorMsg)
      throw new Error(errorMsg)
    }
    console.log(`[API] Successfully fetched ${response.data.data.length} reviews`)
    return response.data
  } catch (error) {
    const errorMsg = error.response?.data?.message || 
                    'Failed to load reviews. Please check your connection and try again.'
    console.error('[API] getMyReviews exception:', errorMsg, error)
    throw new Error(errorMsg)
  }
}

export const getBookReviews = async (userId, bookId) => {
  try {
    console.log(`[API] Fetching reviews for book ID: ${bookId}`)
    const response = await api.post(`me/reviews/${bookId}`, {
      user_id: userId,
    })
    
    if (response.data.status !== 'success') {
      const errorMsg = response.data.message || 'Failed to load book reviews. Please try again.'
      console.error('[API] getBookReviews error:', errorMsg)
      throw new Error(errorMsg)
    }

    console.log(`[API] Successfully fetched ${response.data.data.length} reviews for book ID: ${bookId}`)
    return response.data
  } catch (error) {
    const errorMsg = error.response?.data?.message || 
                    'Failed to load book reviews. The book may not exist or you may not have access.'
    console.error('[API] getBookReviews exception:', errorMsg, error)
    throw new Error(errorMsg)
  }
}

export const createReview = async (userId, bookId, rate, review) => {
  try {
    console.log(`[API] Creating review for book ID: ${bookId}`)
    const response = await api.post(`/me/reviews/${bookId}/new`, {
      user_id: userId,
      rate: rate,
      text: review
    })
    
    if (response.data.status !== 'success') {
      const errorMsg = response.data.message || 'Failed to create review. Please try again.'
      console.error('[API] createReview error:', errorMsg)
      throw new Error(errorMsg)
    }
    return response.data
  } catch (error) {
    const errorMsg = error.response?.data?.message || 
                    'Failed to create review. Please check your connection and try again.'
    console.error('[API] createReview exception:', errorMsg, error)
    throw new Error(errorMsg)
  }
}

export const updateReview = async (userId, bookId, rate, review) => {
  try {
    console.log(`[API] Updating review for book ID: ${bookId}`)
    const response = await api.put(`/me/reviews/${bookId}`, {
      user_id: userId,
      rate: rate,
      text: review
    })
    
    if (response.data.status !== 'success') {
      const errorMsg = response.data.message || 'Failed to update review. Please try again.'
      console.error('[API] updateReview error:', errorMsg)
      throw new Error(errorMsg)
    }

    console.log(`[API] Review updated successfully for book ID: ${bookId}`)
    return response.data
  } catch (error) {
    const errorMsg = error.response?.data?.message || 
                    'Failed to update review. Please check your connection and try again.'
    console.error('[API] updateReview exception:', errorMsg, error)
    throw new Error(errorMsg)
  }
}

export const deleteReview = async (userId, bookId) => {
  try {
    console.log(`[API] Deleting review for book ID: ${bookId}`)
    const response = await api.delete(`/me/reviews/${bookId}`, {
      data: {
        user_id: userId
      }
    })
    
    if (response.data.status !== 'success') {
      const errorMsg = response.data.message || 'Failed to delete review. Please try again.'
      console.error('[API] deleteReview error:', errorMsg)
      throw new Error(errorMsg)
    }

    console.log(`[API] Review deleted successfully for book ID: ${bookId}`)
    return response.data
  } catch (error) {
    const errorMsg = error.response?.data?.message || 
                    'Failed to delete review. Please check your connection and try again.'
    console.error('[API] deleteReview exception:', errorMsg, error)
    throw new Error(errorMsg)
  }
}
